### PR TITLE
greybus: subsys: remove deferred init, allow usage of multiple cports on same bundle

### PR DESCRIPTION
**Pull request by vaishnav98**
Hi @cfriedt ,
These changes make use of priority level to fix the device initialization priority and hope that we can remove the deferred init setup using hardcoded priority for interface, bundle, control... devices, tested on the latest setup with SPI,I2C and GPIO and all seems to work well, the main aim of these changes is to allow multiple cports of same bundle class within the same bundle which makes fetching all of the related controllers[Eg: SPI,I2C, GPIO on a mikroBUS port] easier at the host kernel

### CODE REVIEW

File: subsys/greybus/platform/bundle.c [hunk 1]
```diff
@@ -12,6 +12,8 @@ LOG_MODULE_REGISTER(greybus_platform_bundle);
 
 #include <greybus/platform.h>
 
+#define GREYBUS_BUNDLE_DEV_PRIORITY 52
```
Comment (cfriedt) [inline subsys/greybus/platform/bundle.c:15]: Let's move this to `include/greybus/platform.h`

File: subsys/greybus/platform/control.c [hunk 2]
```diff
@@ -11,6 +11,8 @@ LOG_MODULE_REGISTER(greybus_platform_control);
 
 #include <greybus/platform.h>
 
+#define GREYBUS_CONTROL_DEV_PRIORITY 53
```
Comment (cfriedt) [inline subsys/greybus/platform/control.c:14]: Let's move this to `include/greybus/platform.h`

File: subsys/greybus/platform/gpio.c [hunk 3]
```diff
@@ -16,6 +16,8 @@ LOG_MODULE_REGISTER(greybus_platform_gpio_control);
 #include "../gpio-gb.h"
 #include "transport.h"
 
+#define GREYBUS_GPIO_DEV_PRIORITY 53
```
Comment (cfriedt) [inline subsys/greybus/platform/gpio.c:19]: Let's move this to `include/greybus/platform.h`

File: subsys/greybus/platform/i2c.c [hunk 4]
```diff
@@ -16,6 +16,8 @@ LOG_MODULE_REGISTER(greybus_platform_i2c_control);
 #include "../i2c-gb.h"
 #include "transport.h"
 
+#define GREYBUS_I2C_DEV_PRIORITY 53
```
Comment (cfriedt) [inline subsys/greybus/platform/i2c.c:19]: Let's move this to `include/greybus/platform.h`

File: subsys/greybus/platform/interface.c [hunk 5]
```diff
@@ -12,6 +12,8 @@ LOG_MODULE_REGISTER(greybus_platform_interface);
 
 #include <greybus/platform.h>
 
+#define GREYBUS_INTERFACE_DEV_PRIORITY 50
```
Comment (cfriedt) [inline subsys/greybus/platform/interface.c:15]: Let's move this to `include/greybus/platform.h`

File: subsys/greybus/platform/spi.c [hunk 6]
```diff
@@ -16,6 +16,8 @@ LOG_MODULE_REGISTER(greybus_platform_spi_control);
 #include "../spi-gb.h"
 #include "transport.h"
 
+#define GREYBUS_SPI_DEV_PRIORITY 53
```
Comment (cfriedt) [inline subsys/greybus/platform/spi.c:19]: Let's move this to `include/greybus/platform.h`

File: subsys/greybus/platform/string.c [hunk 7]
```diff
@@ -10,6 +10,8 @@ LOG_MODULE_REGISTER(greybus_platform_string);
 
 #include <greybus/platform.h>
 
+#define GREYBUS_STRING_PRIORITY 51
```
Comment (cfriedt) [inline subsys/greybus/platform/string.c:13]: Let's move this to `include/greybus/platform.h`

**Comment by cfriedt** [comment]
@vaishnav98 

Looks great aside from some minor change requests.

I'm also thinking it might make sense to namespace these defines with possibly a `Z_` prefix because they are RTOS-dependent.

I did a quick `grep -R "#define.*_PRIO"` and I thought it might be possible to reuse existing generic priority levels as a reference. In particular, 

```
build/zephyr/include/generated/autoconf.h:#define CONFIG_KERNEL_INIT_PRIORITY_OBJECTS 30
build/zephyr/include/generated/autoconf.h:#define CONFIG_KERNEL_INIT_PRIORITY_DEFAULT 40
build/zephyr/include/generated/autoconf.h:#define CONFIG_KERNEL_INIT_PRIORITY_DEVICE 50
build/zephyr/include/generated/autoconf.h:#define CONFIG_APPLICATION_INIT_PRIORITY 90
```
If we look at a possible hierarchy of dependencies

string descriptors: (no dependencies)
interface descriptors: string
bundle descriptors: interface (?)
cport descriptors: bundle, device

I think all of these things can happen after `CONFIG_KERNEL_INIT_PRIORITY_DEVICE`, but they should probably happen before `CONFIG_APPLICATION_INIT_PRIORITY`.

Do you agree with that assessment? 

I'm thinking the defines below might make sense. It might also not be necessary to give the control cport special consideration, so they may be able to be lumped into generic categories.

```cpp
#define Z_GREYBUS_SERVICE_PRIORITY (CONFIG_APPLICATION_INIT_PRIORITY - 1)
#define Z_GREYBUS_CPORT_PRIORITY (GREYBUS_SERVICE_PRIORITY - 1)
#define Z_GREYBUS_BUNDLE_PRIORITY (GREYBUS_CPORT_PRIORITY - 1)
#define Z_GREYBUS_INTERFACE_PRIORITY (GREYBUS_BUNDLE_PRIORITY - 1)
#define Z_GREYBUS_INTERFACE_PRIORITY (GREYBUS_BUNDLE_PRIORITY - 1)
```

**Comment by vaishnav98** [comment]
> @vaishnav98
> 
> Looks great aside from some minor change requests.
> 
> I'm also thinking it might make sense to namespace these defines with possibly a `Z_` prefix because they are RTOS-dependent.
> 
> I did a quick `grep -R "#define.*_PRIO"` and I thought it might be possible to reuse existing generic priority levels as a reference. In particular,
> 
> ```
> build/zephyr/include/generated/autoconf.h:#define CONFIG_KERNEL_INIT_PRIORITY_OBJECTS 30
> build/zephyr/include/generated/autoconf.h:#define CONFIG_KERNEL_INIT_PRIORITY_DEFAULT 40
> build/zephyr/include/generated/autoconf.h:#define CONFIG_KERNEL_INIT_PRIORITY_DEVICE 50
> build/zephyr/include/generated/autoconf.h:#define CONFIG_APPLICATION_INIT_PRIORITY 90
> ```
> 
> If we look at a possible hierarchy of dependencies
> 
> string descriptors: (no dependencies)
> interface descriptors: string
> bundle descriptors: interface (?)
> cport descriptors: bundle, device
> 
> I think all of these things can happen after `CONFIG_KERNEL_INIT_PRIORITY_DEVICE`, but they should probably happen before `CONFIG_APPLICATION_INIT_PRIORITY`.
> 
> Do you agree with that assessment?
> 
> I'm thinking the defines below might make sense. It might also not be necessary to give the control cport special consideration, so they may be able to be lumped into generic categories.
> 
> ```c++
> #define Z_GREYBUS_SERVICE_PRIORITY (CONFIG_APPLICATION_INIT_PRIORITY - 1)
> #define Z_GREYBUS_CPORT_PRIORITY (GREYBUS_SERVICE_PRIORITY - 1)
> #define Z_GREYBUS_BUNDLE_PRIORITY (GREYBUS_CPORT_PRIORITY - 1)
> #define Z_GREYBUS_INTERFACE_PRIORITY (GREYBUS_BUNDLE_PRIORITY - 1)
> #define Z_GREYBUS_INTERFACE_PRIORITY (GREYBUS_BUNDLE_PRIORITY - 1)
> ```
@cfriedt 
Your above suggestions sound great, I have addressed them in the latest push, using an existing generic priority level as reference makes things much neater and yes we don't need to give control cport special consideration, tested everything again on sensortag

**Comment by vaishnav98** [comment]
Hi @cfriedt , converting to draft temporarily as facing some issues with I2C, will test thoroughly and request for your review again

**Comment by cfriedt** [comment]
@vaishnav98

A few issues I've found after trying a few things with your change.
1. the priority must reduce to a constant integer (not a function of some other macro), so I suggest we use Kconfig to define those constants. It's because Zephyr does some elf section manipulation (see below [1]), so macro expressions will not work
1. basically all of the entities under the bus depend on the bus "device" to be present because they need to access the bus->api and add themselves to the bus (for the purpose of generating a manifest), but we don't want to complete the entire `greybus_service_init()` when the bus is initially probed. Maybe it would make sense to have the bus just do a very basic initialization of itself when probed, and then later from the application we could call greybus_service_start() which would do all of the network init and so on.
1. I think I had the priority of the service incorrect originally, so it should look more like [2], where N is equal to `CONFIG_APPLICATION_INIT_PRIORITY`, but it would definitely require splitting up `greybus_service_init()`

Thanks for your work - I appreciate that you're putting effort into this - it really is a bit of a tricky problem.

[1]
```bash
hyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL((90-1)-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(bundle.c.obj)' being placed in section `.device_POST_KERNEL((90-1)-1)'
/home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL(((((90-1)-1)-1)-1)-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(bus.c.obj)' being placed in section `.device_POST_KERNEL(((((90-1)-1)-1)-1)-1)'
/home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL(90-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(control.c.obj)' being placed in section `.device_POST_KERNEL(90-1)'
/home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL(((90-1)-1)-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(interface.c.obj)' being placed in section `.device_POST_KERNEL(((90-1)-1)-1)'
/home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL((((90-1)-1)-1)-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(string.c.obj)' being placed in section `.device_POST_KERNEL((((90-1)-1)-1)-1)'
/home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL(90-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(gpio.c.obj)' being placed in section `.device_POST_KERNEL(90-1)'
/home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL(90-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(i2c.c.obj)' being placed in section `.device_POST_KERNEL(90-1)'
/home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL(90-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(spi.c.obj)' being placed in section `.device_POST_KERNEL(90-1)'
/home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: Undefined initialization levels used.
collect2: error: ld returned 1 exit status
%
```

[2]
```cpp
#define Z_GREYBUS_SERVICE_PRIORITY N-5
#define Z_GREYBUS_STRING_PRIORITY N-4
#define Z_GREYBUS_INTERFACE_PRIORITY N-3
#define Z_GREYBUS_BUNDLE_PRIORITY N-2
#define Z_GREYBUS_CPORT_PRIORITY N-1
```

**Comment by vaishnav98** [comment]
> @vaishnav98
> 
> A few issues I've found after trying a few things with your change.
> 
> 1. the priority must reduce to a constant integer (not a function of some other macro), so I suggest we use Kconfig to define those constants. It's because Zephyr does some elf section manipulation (see below [1]), so macro expressions will not work
> 2. basically all of the entities under the bus depend on the bus "device" to be present because they need to access the bus->api and add themselves to the bus (for the purpose of generating a manifest), but we don't want to complete the entire `greybus_service_init()` when the bus is initially probed. Maybe it would make sense to have the bus just do a very basic initialization of itself when probed, and then later from the application we could call a new function like `greybus_service_start()` which would do all of the network init and so on.
> 3. I think I had the priority of the service incorrect originally, so it should look more like [2], where N is equal to `CONFIG_APPLICATION_INIT_PRIORITY`, but it would definitely require splitting up `greybus_service_init()`
> 
> Thanks for your work - I appreciate that you're putting effort into this - it really is a bit of a tricky problem.
> 
> [1]
> 
> ```shell
> hyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL((90-1)-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(bundle.c.obj)' being placed in section `.device_POST_KERNEL((90-1)-1)'
> /home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL(((((90-1)-1)-1)-1)-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(bus.c.obj)' being placed in section `.device_POST_KERNEL(((((90-1)-1)-1)-1)-1)'
> /home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL(90-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(control.c.obj)' being placed in section `.device_POST_KERNEL(90-1)'
> /home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL(((90-1)-1)-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(interface.c.obj)' being placed in section `.device_POST_KERNEL(((90-1)-1)-1)'
> /home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL((((90-1)-1)-1)-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(string.c.obj)' being placed in section `.device_POST_KERNEL((((90-1)-1)-1)-1)'
> /home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL(90-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(gpio.c.obj)' being placed in section `.device_POST_KERNEL(90-1)'
> /home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL(90-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(i2c.c.obj)' being placed in section `.device_POST_KERNEL(90-1)'
> /home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: warning: orphan section `.device_POST_KERNEL(90-1)' from `zephyr/subsys/greybus/libsubsys__greybus.a(spi.c.obj)' being placed in section `.device_POST_KERNEL(90-1)'
> /home/cfriedt/zephyr-sdk-0.11.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: Undefined initialization levels used.
> collect2: error: ld returned 1 exit status
> %
> ```
> 
> [2]
> 
> ```c++
> #define Z_GREYBUS_SERVICE_PRIORITY N-5
> #define Z_GREYBUS_STRING_PRIORITY N-4
> #define Z_GREYBUS_INTERFACE_PRIORITY N-3
> #define Z_GREYBUS_BUNDLE_PRIORITY N-2
> #define Z_GREYBUS_CPORT_PRIORITY N-1
> ```

Thank you @cfriedt , yes we need to keep the priority to constants and expressions are leading to linker error as you were facing, will move these to KConfig to define the priorities as per the scheme [2] above, also moving the multiple cports to the same bundle seems to cause some problems, so shall I move the overlay change temporarily from this PR and submit separately?, the issue faced is with I2C, SPI and GPIO work together well without any issues so confused where the issue lies.

File: include/greybus/platform.h [hunk 8]
```diff
@@ -9,6 +9,13 @@
 
 #include <greybus/manifecto/manifest.h>
 
+/* To ensure greybus device initialization order */
```
Comment (cfriedt) [inline include/greybus/platform.h:12]: Let's convert these to constants defined inside of Kconfig and remove this section.

**Comment by cfriedt** [comment]
- [ ] Uncomment the `SYS_INIT` macro in service.c it and convert it to
```cpp
SYS_INIT(greybus_service_init, APPLICATION,
         CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```
- [ ] Add this in `greybus_service_init()` add this right after variables are declared:
```cpp
    if (xport != NULL) {
    	return -EALREADY;
    }
```
- [ ] remove the manual call to `greybus_service_init()` in the main.c files in the `greybus/net` and `greybus/uart` samples.
- [ ] Clean up includes in the above main.c files to only include `<zephyr.h>`. Remove the `DEBUG` macro.
- [ ] Make `greybus_service_init()` static.

**Comment by cfriedt** [comment]
Just learned that the original NuttX code used one device per bundle, 
https://github.com/projectara/nuttx/blob/master/nuttx/include/nuttx/greybus/greybus.h#L109
Relevant problematic code is here:
https://github.com/projectara/nuttx/blob/master/nuttx/drivers/greybus/i2c.c#L151
So it likely won't work currently to add multiple cports into a single bundle

**Comment by vaishnav98** [comment]
Closing this PR, reopened new PR @ https://github.com/cfriedt/greybus-for-zephyr/pull/50 with all the review comments addressed

**Pull request closed without merging**